### PR TITLE
stricter lints

### PR DESCRIPTION
**Pull request by nico-famedly**
I guess the only controversial change is the runInRoot stuff, but I don't see, why we would ever want to await those. The few cases, where we do, seem to fit better when we just await them directly?

### CODE REVIEW

**Review by techno-disaster — COMMENTED** [review]
Mostly LGTM, can you explain why some of the functions don't run in root anymore? `updateSessionUsage` for example

File: analysis_options.yaml [hunk 1]
```diff
@@ -2,19 +2,24 @@ include: package:lints/recommended.yaml
 
 linter:
   rules:
-    camel_case_types: true
+    always_declare_return_types: true
```
Comment (krille-chan) [inline analysis_options.yaml:5]: Do you have some infos how you have decided for this lint rules? Are they from a source or have you checked every lint rule by yourself if they make sense? I would prefer to have a reason for any point where we differ from the default

File: lib/encryption/encryption.dart [hunk 2]
```diff
@@ -100,11 +100,12 @@ class Encryption {
 
   void handleDeviceOneTimeKeysCount(
       Map<String, int>? countJson, List<String>? unusedFallbackKeyTypes) {
-    runInRoot(() => olmManager.handleDeviceOneTimeKeysCount(
+    runInRoot(() async => olmManager.handleDeviceOneTimeKeysCount(
```
Comment (krille-chan) [inline lib/encryption/encryption.dart:103]: I don't get why this makes a difference 🤔 the method `handleDeviceOneTimeKeysCount` is void and not `Future<void>` anyway and not even async so we are not awaiting this anyway

Comment (krille-chan) [reply lib/encryption/encryption.dart:103]: Probably a lint rule enforces this?

File: analysis_options.yaml (hunk 1 above)
Comment (nico-famedly) [reply analysis_options.yaml:5]: I checked each of those rules one by one depending on what value they provide. I don't think that needs to be repeated in the config file, since the reasoning is pretty much the same as in their description, however to repeat it here:

- avoid_bool_literals_in_conditional_expressions is about cases like `condition ? true : boolExpression`, which are just a more verbose way to say `condition || boolExpression`
- cancel_subscriptions: Is about not leaking scream subscriptions, which would cause performance degradations over time or memory leaks
- discarded_futures: is like unawaited futures, but in functions which have not been marked as async. We tend to make this mistake quite often, that we don't await something and don't document why. This forces a linter annotation at least for cases where we do want that.
- no_adjacent_strings_in_list: Is about cases like `['a' 'b', 'c']`. Someone forgot the comma there. If that was intentional, they can pack it in one string literal or use a + to concatenate.
- test_types_in_equals: is a bit niche, so we could probably drop it. It is just on how to implement operator== correctly.
- throw_in_finally: Is a warning that throwing in a finally block is usually unintentional and could lead to weird program execution. Generally you want the throw statement outside of the finally block.
- unnecessary_statements: Just warns on statements, that return a value, that is never assigned and that have no side effect. It doesn't warn on function calls, that just return a value, but if you for example add `1+1` and don't assign that to a variable, that was probably a mistake.

In general I only enabled linter flags, that seem to have no drawbacks, but feel free to point out any that do. We diverge here from the default by being stricter than the default lints. We do that, because we want to avoid more bugs and this has pretty much no effect on existing code apart from where it was possibly already buggy to begin with.

File: lib/encryption/encryption.dart (hunk 2 above)
Comment (nico-famedly) [reply lib/encryption/encryption.dart:103]: Yes, that is the one drawback I found from the discarded_futures. It does warn you when using a Future in a closure, but the closure is not marked as async. However I do think the value it provides elsewhere is beneficial.

File: analysis_options.yaml (hunk 1 above)
Comment (krille-chan) [reply analysis_options.yaml:5]: Are you sure that all of those rules are not in the list of default ones? Might sound reasonable to not add an argumentation for every single rule but a comment above them to explain why this was added and maybe explain one or two could help in the future to understand this

File: lib/encryption/encryption.dart (hunk 2 above)
Comment (krille-chan) [reply lib/encryption/encryption.dart:103]: Okay sounds fair

File: analysis_options.yaml (hunk 1 above)
Comment (nico-famedly) [reply analysis_options.yaml:5]: I am pretty sure they are not in the default set. In fact I removed a rule, that was in the default set, that we had explicitly enabled (`avoid_function_literals_in_foreach_calls`).

Can you give me an example of what that comment would look like? I think it is obvious, that linter rules are enabled to prevent bugs and each of those linter rules has a better explanation in the dart docs, that I would prefer not to duplicate. And so far we never explained, why we enable linter rules either and we do have quite a few of them. We only document, why we disable some. As such, I have no idea what the comment would say...

Basically what I always did is, if I didn't understand a linter rule, I went to https://dart.dev/tools/linter-rules/avoid_function_literals_in_foreach_calls to figure out what it does and why one would want to use it. That seems to be more beneficial to me, than duplicating that information in a worse way.

Comment (krille-chan) [reply analysis_options.yaml:5]: Hmm something like:

# Added more strict linter rules than in default for better code quality.
# We picked "avoid_bool_literals_in_conditional_expressions", "no_adjacent_strings_in_list" and
# "unnecessary_statements" for better readability and "cancel_subscriptions", "discarded_futures" to
# avoid performance leaks.

Or maybe even group the lints by topics like "Code readability", "Performance" and so on

**Review by krille-chan — DISMISSED** [review]

File: analysis_options.yaml (hunk 1 above)
Comment (krille-chan) [reply analysis_options.yaml:5]: However, you will do it right. Have approved it :)
